src/lib/reasoners/bitv.ml

```diff
@@ -64,16 +64,117 @@ let pp_sort ppf = function
   | B -> Format.fprintf ppf "b"
   | C -> Format.fprintf ppf "c"
 
-let compare_sort s1 s2 =
-  match s1, s2 with
-  | A, A | B, B | C, C -> 0
-  | A, (B | C) | B, C -> -1
-  | C, (A | B) | B, A -> 1
-
-type tvar = { var : int ; sorte : sort_var }
-
-let pp_tvar ppf { var; sorte } =
-  Format.fprintf ppf "%a_%d" pp_sort sorte var
+(** The [tvar] and [defn] type are used to implement Tarjan's Union-Find
+    algorithm with path compression but not union-by-rank.
```

Comment:
There are two reasons, mostly:

 - I didn't want to add extra complexity to the implementation — there would be some subtleties to take care of as you note (mostly you need to keep `Cte` when one of the members is a `Cte`), but union-by-rank would work otherwise and yield better complexity.
 - The size of the problems here are relatively small (for an equality on `(_ BitVec n)` there is at most one variable per bit, if I am not mistaken) that I'm not sure it'd matter too much in practice.

That said, it wouldn't hurt to have union-by-rank here, I can make an issue (however I have since realized a performance bug in the current representation of bitvectors which is that `x @ y = 0b0000000000` will use 2 variables but `x @ y = 0b1010101010` will use 10 and that I want to fix first, as it will have to change the representation of the solver a bit).

Reply:
I agree, we don't need a too complicated solution here. Let's merge.